cpp/src/arrow/array/data.cc

```diff
@@ -224,6 +224,41 @@ int64_t ArrayData::ComputeLogicalNullCount() const {
   return ArraySpan(*this).ComputeLogicalNullCount();
 }
 
+DeviceAllocationType ArrayData::device_type() const {
+  // we're using 0 as a sentinel value for NOT YET ASSIGNED
+  // there is explicitly no constant DeviceAllocationType to represent
+  // the "UNASSIGNED" case as it is invalid for data to not have an
+  // assigned device type. If it's still 0 at the end, then we return
+  // CPU as the allocation device type
+  int type = 0;
+  for (const auto& buf : buffers) {
+    if (!buf) continue;
+    if (type == 0) {
+      type = static_cast<int>(buf->device_type());
+    } else {
+      DCHECK_EQ(type, static_cast<int>(buf->device_type()));
```

Comment:
We don't currently disallow it on creation. We don't do any checking of the buffer device types on ArrayData or Array creation currently. It feels wrong to do so in a non-debug context but it might make sense to add debug checks to confirm the device types of all the buffers match in the constructor. Thoughts?

Reply:
>  it might make sense to add debug checks to confirm the device types of all the buffers match in the constructor

I think that would make sense yes, if we don't want to support that (I don't know if there would actually be use cases for allowing it, in which case I would rather expect a normal error here when trying to export it, saying it requires a single device type)

Reply:
Added the debug checks in the ArrayData constructors